analyze_lua_directives.go

```diff
@@ -0,0 +1,261 @@
+//   - whether or not it is a block directive
+//   - whether this is a flag (takes one argument that's either "on" or "off")
+//   - which contexts it's allowed to be in
+
+package crossplane
+
+//nolint:gochecknoglobals
+var moduleLuaDirectives = map[string][]uint{
+	"access_by_lua": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"access_by_lua_block": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"access_by_lua_file": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"access_by_lua_no_postpone": {
+		ngxHTTPMainConf | ngxConfFlag,
+	},
+	"balancer_by_lua_block": {
+		ngxHTTPUpsConf | ngxConfTake1,
+	},
+	"balancer_by_lua_file": {
+		ngxHTTPUpsConf | ngxConfTake1,
+	},
+	"body_filter_by_lua": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"body_filter_by_lua_block": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"body_filter_by_lua_file": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"content_by_lua": {
+		ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"content_by_lua_block": {
+		ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"content_by_lua_file": {
+		ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"exit_worker_by_lua_block": {
+		ngxHTTPMainConf | ngxConfTake1,
+	},
+	"exit_worker_by_lua_file": {
+		ngxHTTPMainConf | ngxConfTake1,
+	},
+	"header_filter_by_lua": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"header_filter_by_lua_block": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"header_filter_by_lua_file": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"init_by_lua": {
+		ngxHTTPMainConf | ngxConfTake1,
+	},
+	"init_by_lua_block": {
+		ngxHTTPMainConf | ngxConfTake1,
+	},
+	"init_by_lua_file": {
+		ngxHTTPMainConf | ngxConfTake1,
+	},
+	"init_worker_by_lua": {
+		ngxHTTPMainConf | ngxConfTake1,
+	},
+	"init_worker_by_lua_block": {
+		ngxHTTPMainConf | ngxConfTake1,
+	},
+	"init_worker_by_lua_file": {
+		ngxHTTPMainConf | ngxConfTake1,
+	},
+	"log_by_lua": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"log_by_lua_block": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"log_by_lua_file": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"lua_capture_error_log": {
+		ngxHTTPMainConf | ngxConfTake1,
+	},
+	"lua_check_client_abort": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfFlag,
+	},
+	"lua_code_cache": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfFlag,
+	},
+	"lua_fake_shm": {
```

Comment:
This directive is not introduced in the Lua docs https://github.com/openresty/lua-nginx-module?tab=readme-ov-file#directives

Reply:
It is a directive in [/lua-nginx-module/t/data/fake-shm-module/ngx_http_lua_fake_shm_module.c](https://github.com/openresty/lua-nginx-module/blob/master/t/data/fake-shm-module/ngx_http_lua_fake_shm_module.c). I guess they want to drop the support after a period so it is undocumented

Reply:
https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_module.c 
we reference to this file

Reply:
Both of these two files are in the [lua git repo](https://github.com/openresty/lua-nginx-module/tree/master). And it seems there is still a [test case](https://github.com/openresty/lua-nginx-module/blob/master/t/149-hup-fake-shm-zone.t) for `lua_fake_shm`. The last commit for `lua_fake_shm ` is 9 years ago. I think this directive is likely an outdated Lua directive(not pretty sure)?

I just run a script and it seems [ngx_http_lua_module.c](https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_module.c) covers all directives except `lua_fake_shm`. Currently the generator can rely on it, but I'm afraid they may split `ngx_http_lua_module.c` or have some directive elsewhere. What about generate for all directives on [lua git repo](https://github.com/openresty/lua-nginx-module/tree/master), then use the [lua doc](https://github.com/openresty/lua-nginx-module?tab=readme-ov-file#directives) as a filter? I can add it in next MR